File: src/main/java/com/brandonitaly/bedrockskins/client/BedrockPlayerModel.java
package com.brandonitaly.bedrockskins.client;

import com.brandonitaly.bedrockskins.bedrock.BedrockBone;
import com.brandonitaly.bedrockskins.bedrock.BedrockGeometry;
import net.minecraft.client.model.*;
import net.minecraft.client.model.geom.ModelPart;
import net.minecraft.client.model.geom.PartNames;
import net.minecraft.client.model.geom.PartPose;
import net.minecraft.client.model.geom.builders.CubeDeformation;
import net.minecraft.client.model.geom.builders.CubeListBuilder;
import net.minecraft.client.model.geom.builders.LayerDefinition;
import net.minecraft.client.model.geom.builders.MeshDefinition;
import net.minecraft.client.model.geom.builders.PartDefinition;
import net.minecraft.client.model.player.PlayerModel;
import net.minecraft.client.renderer.entity.state.AvatarRenderState;
import java.util.*;

public class BedrockPlayerModel extends PlayerModel {
    public final ModelPart root;
    public final Map<String, ModelPart> partsMap;
    public final Map<String, PartTransform> defaultTransforms;
    public float armorYOffset = 0f;
    public float capeYOffset = 0f;
    public float upperArmorYOffset = 0f;
    private final boolean animationArmsOutFront;
    private final boolean animationStationaryLegs;

    public BedrockPlayerModel(ModelPart root, boolean thinArms, Map<String, ModelPart> partsMap, Map<String, PartTransform> defaultTransforms, boolean animationArmsOutFront, boolean animationStationaryLegs) {
        super(root, thinArms);
        this.root = root;
        this.partsMap = Collections.unmodifiableMap(new HashMap<>(partsMap));
        this.defaultTransforms = Collections.unmodifiableMap(new HashMap<>(defaultTransforms));
        this.animationArmsOutFront = animationArmsOutFront;
        this.animationStationaryLegs = animationStationaryLegs;
    }

    public static class PartTransform {
        public final float x, y, z, pitch, yaw, roll;
        public PartTransform(float x, float y, float z, float pitch, float yaw, float roll) { this.x = x; this.y = y; this.z = z; this.pitch = pitch; this.yaw = yaw; this.roll = roll; }
    }

    public static BedrockPlayerModel create(BedrockGeometry geometry, boolean thinArms) {
        validateAndPatchGeometry(geometry);
        BuildRootResult res = buildRoot(geometry);
        return new BedrockPlayerModel(res.root, thinArms, res.parts, res.defaults, geometry.getAnimationArmsOutFront() != null ? geometry.getAnimationArmsOutFront() : false, geometry.getAnimationStationaryLegs() != null ? geometry.getAnimationStationaryLegs() : false);
    }

    private static void validateAndPatchGeometry(BedrockGeometry geometry) {
        Map<String, String> requiredBones = new HashMap<>();
        requiredBones.put("head", "body");
        requiredBones.put("hat", "head");
        requiredBones.put("body", null);
        requiredBones.put("jacket", "body");
        requiredBones.put("leftArm", "body");
        requiredBones.put("leftSleeve", "leftArm");
        requiredBones.put("rightArm", "body");
        requiredBones.put("rightSleeve", "rightArm");
        requiredBones.put("leftLeg", "body");
        requiredBones.put("leftPants", "leftLeg");
        requiredBones.put("rightLeg", "body");
        requiredBones.put("rightPants", "rightLeg");

        List<BedrockBone> currentBones = geometry.getBones() != null ? new ArrayList<>(geometry.getBones()) : new ArrayList<>();
        Set<String> existingBoneNames = new HashSet<>();
        for (BedrockBone b : currentBones) if (b.getName() != null) existingBoneNames.add(b.getName().toLowerCase());

        List<BedrockBone> newBones = new ArrayList<>();
        for (Map.Entry<String, String> e : requiredBones.entrySet()) {
            String boneName = e.getKey();
            String parentName = e.getValue();
            boolean isMissing = !existingBoneNames.contains(boneName.toLowerCase()) && !existingBoneNames.contains(mapBoneName(boneName).toLowerCase());
            if (isMissing) {
                // System.out.println("BedrockPlayerModel: Patching geometry - adding missing bone: " + boneName);
                BedrockBone b = new BedrockBone();
                b.setName(boneName);
                b.setParent(parentName);
                b.setPivot(Arrays.asList(0f, 0f, 0f));
                b.setRotation(Arrays.asList(0f, 0f, 0f));
                b.setCubes(Collections.emptyList());
                b.setMirror(false);
                newBones.add(b);
            }
        }
        if (!newBones.isEmpty()) {
            List<BedrockBone> combined = new ArrayList<>(currentBones);
            combined.addAll(newBones);
            geometry.setBones(combined);
        }
    }

    private static class BuildRootResult {
        final ModelPart root;
        final Map<String, ModelPart> parts;
        final Map<String, PartTransform> defaults;
        BuildRootResult(ModelPart root, Map<String, ModelPart> parts, Map<String, PartTransform> defaults) { this.root = root; this.parts = parts; this.defaults = defaults; }
    }

    private static BuildRootResult buildRoot(BedrockGeometry geometry) {
        MeshDefinition modelData = new MeshDefinition();
        PartDefinition rootData = modelData.getRoot();

        Map<String, BedrockBone> boneMap = new HashMap<>();
        if (geometry.getBones() != null) for (BedrockBone b : geometry.getBones()) boneMap.put(b.getName(), b);
        Set<String> processedBones = new HashSet<>();

        if (geometry.getBones() != null) {
            for (BedrockBone b : geometry.getBones()) {
                addBoneRecursively(b.getName(), boneMap, processedBones);
            }
        }

        Map<String, PartDefinition> partDataMap = new HashMap<>();
        Set<String> vanillaRootParts = new HashSet<>(Arrays.asList(
                PartNames.HEAD,
                PartNames.BODY,
                PartNames.RIGHT_ARM,
                PartNames.LEFT_ARM,
                PartNames.RIGHT_LEG,
                PartNames.LEFT_LEG
        ));

        List<BedrockBone> bonesToProcess = geometry.getBones() != null ? new ArrayList<>(geometry.getBones()) : new ArrayList<>();
        Map<String, PartTransform> defaultTransforms = new HashMap<>();

        int stuckCounter = 0;

        while (!bonesToProcess.isEmpty()) {
            Iterator<BedrockBone> iterator = bonesToProcess.iterator();
            boolean processedAny = false;
            while (iterator.hasNext()) {
                BedrockBone bone = iterator.next();
                if (bone.getParent() == null || partDataMap.containsKey(bone.getParent())) {
                    PartDefinition parentData = bone.getParent() == null ? rootData : partDataMap.get(bone.getParent());
                    CubeListBuilder builder = CubeListBuilder.create();

                    if (bone.getCubes() != null) {
                        for (var cube : bone.getCubes()) {
                            int u = 0, v = 0;
                            Object uvObj = cube.getUv();
                            if (uvObj instanceof java.util.List) {
                                var list = (java.util.List<?>) uvObj;
                                if (list.size() >= 2) { u = ((Number) list.get(0)).intValue(); v = ((Number) list.get(1)).intValue(); }
                            } else if (uvObj instanceof java.util.Map) {
                                var map = (java.util.Map<?,?>) uvObj;
                                var uvList = (java.util.List<?>) map.get("uv");
                                if (uvList != null && uvList.size() >= 2) { u = ((Number) uvList.get(0)).intValue(); v = ((Number) uvList.get(1)).intValue(); }
                            }
                            float dilation = cube.getInflate() != null ? cube.getInflate() : 0f;
                            boolean isMirrored = cube.getMirror() != null ? cube.getMirror() : (bone.getMirror() != null ? bone.getMirror() : false);
                            float bPx = bone.getPivot() != null && bone.getPivot().size() > 0 ? bone.getPivot().get(0) : 0f;
                            float bPy = bone.getPivot() != null && bone.getPivot().size() > 1 ? bone.getPivot().get(1) : 0f;
                            float bPz = bone.getPivot() != null && bone.getPivot().size() > 2 ? bone.getPivot().get(2) : 0f;
                            float cOx = cube.getOrigin().get(0);
                            float cOy = cube.getOrigin().get(1);
                            float cOz = cube.getOrigin().get(2);
                            float offX = cOx - bPx;
                            float offY = bPy - cOy - cube.getSize().get(1);
                            float offZ = cOz - bPz;
                            builder.mirror(isMirrored).texOffs(u, v).addBox(offX, offY, offZ, cube.getSize().get(0), cube.getSize().get(1), cube.getSize().get(2), new CubeDeformation(dilation));
                        }
                    }

                    float bPx = bone.getPivot() != null && bone.getPivot().size() > 0 ? bone.getPivot().get(0) : 0f;
                    float bPy = bone.getPivot() != null && bone.getPivot().size() > 1 ? bone.getPivot().get(1) : 0f;
                    float bPz = bone.getPivot() != null && bone.getPivot().size() > 2 ? bone.getPivot().get(2) : 0f;
                    float pX = bPx;
                    float pY = 24f - bPy;
                    float pZ = bPz;
                    String vanillaName = mapBoneName(bone.getName());
                    PartDefinition parentForCreation = (bone.getParent() != null && vanillaRootParts.contains(vanillaName)) ? rootData : parentData;
                    if (bone.getParent() != null && parentForCreation != rootData) {
                        BedrockBone parentBone = boneMap.get(bone.getParent());
                        float ppX = parentBone.getPivot() != null && parentBone.getPivot().size() > 0 ? parentBone.getPivot().get(0) : 0f;
                        float ppY = parentBone.getPivot() != null && parentBone.getPivot().size() > 1 ? 24f - parentBone.getPivot().get(1) : 24f;
                        float ppZ = parentBone.getPivot() != null && parentBone.getPivot().size() > 2 ? parentBone.getPivot().get(2) : 0f;
                        pX -= ppX;
                        pY -= ppY;
                        pZ -= ppZ;
                    }
                    float rotX = (float)Math.toRadians(-(bone.getRotation() != null && bone.getRotation().size() > 0 ? bone.getRotation().get(0) : 0f));
                    float rotY = (float)Math.toRadians(-(bone.getRotation() != null && bone.getRotation().size() > 1 ? bone.getRotation().get(1) : 0f));
                    float rotZ = (float)Math.toRadians((bone.getRotation() != null && bone.getRotation().size() > 2 ? bone.getRotation().get(2) : 0f));
                    PartPose transform = PartPose.offsetAndRotation(pX, pY, pZ, rotX, rotY, rotZ);
                    PartDefinition partData = parentForCreation.addOrReplaceChild(vanillaName, builder, transform);
                    partDataMap.put(bone.getName(), partData);
                    defaultTransforms.put(bone.getName(), new PartTransform(pX, pY, pZ, rotX, rotY, rotZ));
                    if (!vanillaName.equals(bone.getName())) {
                        defaultTransforms.put(vanillaName, new PartTransform(pX, pY, pZ, rotX, rotY, rotZ));
                    }
                    iterator.remove();
                    processedAny = true;
                }
            }
            if (!processedAny) {
                stuckCounter++;
                if (stuckCounter > 5) break;
            }
        }

        LayerDefinition texturedModelData = LayerDefinition.create(modelData, geometry.getDescription().getTextureWidth(), geometry.getDescription().getTextureHeight());
        ModelPart rootPart = texturedModelData.bakeRoot();

        Map<String, ModelPart> finalParts = new HashMap<>();
        for (String name : boneMap.keySet()) {
            ModelPart found = findPart(rootPart, name, boneMap);
            if (found != null) finalParts.put(name, found);
        }

        return new BuildRootResult(rootPart, finalParts, defaultTransforms);
    }

    private static ModelPart findPart(ModelPart parent, String boneName, Map<String, BedrockBone> boneMap) {
        BedrockBone bone = boneMap.get(boneName);
        if (bone == null) return null;
        String mappedRootName = mapBoneName(bone.getName());
        if (parent.hasChild(mappedRootName)) return parent.getChild(mappedRootName);
        List<String> path = new ArrayList<>();
        BedrockBone curr = bone;
        while (curr != null) {
            path.add(mapBoneName(curr.getName()));
            String p = curr.getParent();
            curr = p != null ? boneMap.get(p) : null;
        }
        Collections.reverse(path);
        ModelPart currPart = parent;
        for (String segment : path) {
            if (currPart.hasChild(segment)) currPart = currPart.getChild(segment);
            else return null;
        }
        return currPart;
    }

    private static void addBoneRecursively(String boneName, Map<String, BedrockBone> boneMap, Set<String> processedBones) {
        if (!processedBones.add(boneName)) return;
        BedrockBone bone = boneMap.get(boneName);
        if (bone == null) return;
        if (bone.getParent() != null) addBoneRecursively(bone.getParent(), boneMap, processedBones);
    }

    public static String mapBoneName(String name) {
        String lower = name.toLowerCase();
        switch (lower) {
            case "head": return PartNames.HEAD;
            case "hat":
            case "headwear": return PartNames.HAT;
            case "body": return PartNames.BODY;
            case "jacket": return PartNames.JACKET;
            case "rightarm":
            case "right_arm": return PartNames.RIGHT_ARM;
            case "leftarm":
            case "left_arm": return PartNames.LEFT_ARM;
            case "rightleg":
            case "right_leg": return PartNames.RIGHT_LEG;
            case "leftleg":
            case "left_leg": return PartNames.LEFT_LEG;
            case "rightsleeve":
            case "right_sleeve": return "right_sleeve";
            case "leftsleeve":
            case "left_sleeve": return "left_sleeve";
            case "rightpants":
            case "right_pants": return "right_pants";
            case "leftpants":
            case "left_pants": return "left_pants";
            default: return name;
        }
    }

    public void setBedrockPartVisible(String partName, boolean visible) {
        ModelPart p = partsMap.get(partName);
        if (p != null) p.visible = visible;
    }

    @Override
    public void setupAnim(AvatarRenderState state) {
        super.setupAnim(state);
        if (animationArmsOutFront) {
            setArmAngle(partsMap.getOrDefault("rightArm", partsMap.get("right_arm")));
            setArmAngle(partsMap.getOrDefault("leftArm", partsMap.get("left_arm")));
        }
        if (animationStationaryLegs) {
            resetLegAngle("rightLeg", "right_leg");
            resetLegAngle("leftLeg", "left_leg");
        }
    }

    private void setArmAngle(ModelPart part) {
        if (part == null) return;
        part.xRot = -1.5707964f;
        part.yRot = 0f;
        part.zRot = 0f;
    }

    private void resetLegAngle(String key1, String key2) {
        ModelPart leg = partsMap.getOrDefault(key1, partsMap.get(key2));
        PartTransform def = defaultTransforms.getOrDefault(key1, defaultTransforms.get(key2));
        if (leg == null || def == null) return;
        leg.xRot = def.pitch;
        leg.yRot = def.yaw;
        leg.zRot = def.roll;
    }

    public void copyFromVanilla(PlayerModel vanillaModel) {
        // copyRot
        for (var pair : Arrays.asList(new Object[][]{
                {"head", vanillaModel.head},
                {"body", vanillaModel.body},
                {"hat", vanillaModel.hat}
        })) {
            String name = (String) pair[0];
            ModelPart part = (ModelPart) pair[1];
            ModelPart dest = partsMap.get(name);
            if (dest == null) dest = partsMap.get(mapBoneName(name));
            if (dest != null) {
                dest.xRot = part.xRot;
                dest.yRot = part.yRot;
                dest.zRot = part.zRot;
            }
        }

        if (!animationArmsOutFront) {
            ModelPart dest = partsMap.get("rightArm"); if (dest != null) { dest.xRot = vanillaModel.rightArm.xRot; dest.yRot = vanillaModel.rightArm.yRot; dest.zRot = vanillaModel.rightArm.zRot; }
            dest = partsMap.get("leftArm"); if (dest != null) { dest.xRot = vanillaModel.leftArm.xRot; dest.yRot = vanillaModel.leftArm.yRot; dest.zRot = vanillaModel.leftArm.zRot; }
        }
        if (!animationStationaryLegs) {
            ModelPart dest = partsMap.get("rightLeg"); if (dest != null) { dest.xRot = vanillaModel.rightLeg.xRot; dest.yRot = vanillaModel.rightLeg.yRot; dest.zRot = vanillaModel.rightLeg.zRot; }
            dest = partsMap.get("leftLeg"); if (dest != null) { dest.xRot = vanillaModel.leftLeg.xRot; dest.yRot = vanillaModel.leftLeg.yRot; dest.zRot = vanillaModel.leftLeg.zRot; }
        }

        // getPivotY reflectively
        java.util.function.Function<ModelPart, Float> getPivotY = part -> {
            try {
                var field = part.getClass().getDeclaredField("pivotY");
                field.setAccessible(true);
                Object f = field.get(part);
                if (f instanceof Number) return ((Number)f).floatValue();
                return 0f;
            } catch (Exception ex) {
                try {
                    var field = net.minecraft.client.model.geom.ModelPart.class.getDeclaredField("pivotY");
                    field.setAccessible(true);
                    return field.getFloat(part);
                } catch (Exception e) {
                    return 0f;
                }
            }
        };

        try {
            PartTransform bodyTransform = defaultTransforms.getOrDefault("body", defaultTransforms.get("BODY"));
            float bedrockBodyY = bodyTransform != null ? bodyTransform.y : 0f;
            float vanillaBodyPivotY = getPivotY.apply(vanillaModel.body);

            float bedrockHeadY = defaultTransforms.getOrDefault("head", bodyTransform) != null ? defaultTransforms.getOrDefault("head", bodyTransform).y : bedrockBodyY;
            float vanillaHeadPivotY = getPivotY.apply(vanillaModel.head);

            upperArmorYOffset = ((bedrockBodyY + bedrockHeadY) * 0.5f) - ((vanillaBodyPivotY + vanillaHeadPivotY) * 0.5f);
            armorYOffset = upperArmorYOffset;

            PartTransform capeTransform = defaultTransforms.get("cape");
            float bedrockCapeY = capeTransform != null ? capeTransform.y : bedrockBodyY;
            capeYOffset = bedrockCapeY - vanillaBodyPivotY;
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/brandonitaly/bedrockskins/client/BedrockPlayerModel.kt
package com.brandonitaly.bedrockskins.client

import com.brandonitaly.bedrockskins.bedrock.BedrockBone
import com.brandonitaly.bedrockskins.bedrock.BedrockGeometry
import net.minecraft.client.model.*
import net.minecraft.client.render.entity.model.EntityModelPartNames
import net.minecraft.client.render.entity.model.PlayerEntityModel
import net.minecraft.client.render.entity.state.PlayerEntityRenderState
 
class BedrockPlayerModel(
    val root: ModelPart, 
    thinArms: Boolean, 
    val partsMap: Map<String, ModelPart>,
    val defaultTransforms: Map<String, PartTransform>,
    val animationArmsOutFront: Boolean,
    val animationStationaryLegs: Boolean
) : PlayerEntityModel(root, thinArms) {

    data class PartTransform(val x: Float, val y: Float, val z: Float, val pitch: Float, val yaw: Float, val roll: Float)

    // Vertical offsets to apply to armor / cape feature rendering so they line up
    // with Bedrock geometry when that geometry uses non-standard pivots.
    var armorYOffset: Float = 0f
    var capeYOffset: Float = 0f
    var upperArmorYOffset: Float = 0f

    companion object {
        fun create(geometry: BedrockGeometry, thinArms: Boolean): BedrockPlayerModel {
            // Ensure geometry has all required vanilla bones before building
            validateAndPatchGeometry(geometry)

            val (root, parts, defaults) = buildRoot(geometry)
            return BedrockPlayerModel(
                root,
                thinArms,
                parts,
                defaults,
                geometry.animationArmsOutFront ?: false,
                geometry.animationStationaryLegs ?: false
            )
        }

        /**
         * Checks the geometry for missing standard bones (Head, Body, Arms, etc.)
         * and injects dummy bones if they are missing. This prevents the PlayerEntityModel
         * constructor from crashing when looking for required children.
         */
        private fun validateAndPatchGeometry(geometry: BedrockGeometry) {
            val requiredBones = mapOf(
                "head" to "body",
                "hat" to "head",
                "body" to null,
                "jacket" to "body",
                "leftArm" to "body",
                "leftSleeve" to "leftArm",
                "rightArm" to "body",
                "rightSleeve" to "rightArm",
                "leftLeg" to "body",
                "leftPants" to "leftLeg",
                "rightLeg" to "body",
                "rightPants" to "rightLeg"
            )

            // Safely get mutable list of current bones
            val currentBones = geometry.bones?.toMutableList() ?: mutableListOf()
            
            // normalize existing names to lowercase for comparison
            val existingBoneNames = currentBones.mapNotNull { it.name?.lowercase() }.toSet()
            val newBones = mutableListOf<BedrockBone>()

            for ((boneName, parentName) in requiredBones) {
                // Check if the bone exists
                val isMissing = !existingBoneNames.contains(boneName.lowercase()) && 
                                !existingBoneNames.contains(mapBoneName(boneName).lowercase())

                if (isMissing) {
                    println("BedrockPlayerModel: Patching geometry - adding missing bone: $boneName")
                    newBones.add(
                        BedrockBone(
                            name = boneName,
                            parent = parentName,
                            pivot = listOf(0f, 0f, 0f),
                            rotation = listOf(0f, 0f, 0f),
                            cubes = emptyList(),
                            mirror = false,
                            locators = null,
                            inflate = null
                        )
                    )
                }
            }

            // Update the geometry with the new bones
            if (newBones.isNotEmpty()) {
                geometry.bones = currentBones + newBones
            }
        }

        private fun buildRoot(geometry: BedrockGeometry): Triple<ModelPart, Map<String, ModelPart>, Map<String, PartTransform>> {
            val modelData = ModelData()
            val rootData = modelData.root

            val boneMap = geometry.bones?.associateBy { it.name } ?: emptyMap()
            val processedBones = mutableSetOf<String>()

            fun addBone(boneName: String) {
                if (!processedBones.add(boneName)) return
                val bone = boneMap[boneName] ?: return
                bone.parent?.let { addBone(it) }
            }
            geometry.bones?.forEach { addBone(it.name) }

            val partDataMap = mutableMapOf<String, ModelPartData>()
            val vanillaRootParts = setOf(
                EntityModelPartNames.HEAD,
                EntityModelPartNames.BODY,
                EntityModelPartNames.RIGHT_ARM,
                EntityModelPartNames.LEFT_ARM,
                EntityModelPartNames.RIGHT_LEG,
                EntityModelPartNames.LEFT_LEG
            )
            val bonesToProcess = geometry.bones?.toMutableList() ?: mutableListOf()
            val defaultTransforms = mutableMapOf<String, PartTransform>()
            var stuckCounter = 0

            fun extractUV(cube: Any?): Pair<Int, Int> {
                return when (cube) {
                    is List<*> -> if (cube.size >= 2) ((cube[0] as Number).toInt() to (cube[1] as Number).toInt()) else 0 to 0
                    is Map<*, *> -> {
                        val uvList = cube["uv"] as? List<*>
                        if (uvList != null && uvList.size >= 2) ((uvList[0] as Number).toInt() to (uvList[1] as Number).toInt()) else 0 to 0
                    }
                    else -> 0 to 0
                }
            }

            while (bonesToProcess.isNotEmpty()) {
                val iterator = bonesToProcess.iterator()
                var processedAny = false
                while (iterator.hasNext()) {
                    val bone = iterator.next()
                    if (bone.parent == null || partDataMap.containsKey(bone.parent)) {
                        val parentData = if (bone.parent == null) rootData else partDataMap[bone.parent]!!
                        val builder = ModelPartBuilder.create()

                        bone.cubes?.forEach { cube ->
                            val (u, v) = extractUV(cube.uv)
                            val dilation = Dilation(cube.inflate ?: 0f)
                            val isMirrored = cube.mirror ?: bone.mirror ?: false
                            val bPx = bone.pivot?.get(0) ?: 0f
                            val bPy = bone.pivot?.get(1) ?: 0f
                            val bPz = bone.pivot?.get(2) ?: 0f
                            val cOx = cube.origin[0]
                            val cOy = cube.origin[1]
                            val cOz = cube.origin[2]
                            val offX = cOx - bPx
                            val offY = bPy - cOy - cube.size[1]
                            val offZ = cOz - bPz
                            builder.mirrored(isMirrored).uv(u, v).cuboid(
                                offX, offY, offZ,
                                cube.size[0], cube.size[1], cube.size[2],
                                dilation
                            )
                        }

                        val bPx = bone.pivot?.get(0) ?: 0f
                        val bPy = bone.pivot?.get(1) ?: 0f
                        val bPz = bone.pivot?.get(2) ?: 0f
                        var pX = bPx
                        var pY = 24f - bPy
                        var pZ = bPz
                        val vanillaName = mapBoneName(bone.name)
                        val parentForCreation = if (bone.parent != null && vanillaRootParts.contains(vanillaName)) rootData else parentData
                        if (bone.parent != null && parentForCreation !== rootData) {
                            val parentBone = boneMap[bone.parent]!!
                            val ppX = parentBone.pivot?.get(0) ?: 0f
                            val ppY = 24f - (parentBone.pivot?.get(1) ?: 0f)
                            val ppZ = parentBone.pivot?.get(2) ?: 0f
                            pX -= ppX
                            pY -= ppY
                            pZ -= ppZ
                        }
                        val rotX = Math.toRadians(-(bone.rotation?.get(0) ?: 0f).toDouble()).toFloat()
                        val rotY = Math.toRadians(-(bone.rotation?.get(1) ?: 0f).toDouble()).toFloat()
                        val rotZ = Math.toRadians((bone.rotation?.get(2) ?: 0f).toDouble()).toFloat()
                        val transform = ModelTransform.of(pX, pY, pZ, rotX, rotY, rotZ)
                        val partData = parentForCreation.addChild(vanillaName, builder, transform)
                        partDataMap[bone.name] = partData
                        defaultTransforms[bone.name] = PartTransform(pX, pY, pZ, rotX, rotY, rotZ)
                        if (vanillaName != bone.name) {
                            defaultTransforms[vanillaName] = PartTransform(pX, pY, pZ, rotX, rotY, rotZ)
                        }
                        iterator.remove()
                        processedAny = true
                    }
                }
                if (!processedAny) {
                    stuckCounter++
                    if (stuckCounter > 5) break
                }
            }

            val texturedModelData = TexturedModelData.of(modelData, geometry.description.textureWidth, geometry.description.textureHeight)
            val rootPart = texturedModelData.createModel()

            val finalParts = mutableMapOf<String, ModelPart>()
            fun findPart(parent: ModelPart, boneName: String): ModelPart? {
                val bone = boneMap[boneName] ?: return null
                val mappedRootName = mapBoneName(bone.name)
                if (parent.hasChild(mappedRootName)) {
                    return parent.getChild(mappedRootName)
                }
                val path = generateSequence(bone) { b -> b.parent?.let { boneMap[it] } }
                    .map { mapBoneName(it.name) }
                    .toList().asReversed()
                var currPart = parent
                for (segment in path) {
                    if (currPart.hasChild(segment)) {
                        currPart = currPart.getChild(segment)
                    } else {
                        return null
                    }
                }
                return currPart
            }
            boneMap.keys.forEach { name ->
                findPart(rootPart, name)?.let { finalParts[name] = it }
            }
            return Triple(rootPart, finalParts, defaultTransforms)
        }

        fun mapBoneName(name: String): String {
            val lower = name.lowercase()
            return when {
                lower == "head" -> EntityModelPartNames.HEAD // "head"
                lower == "hat" || lower == "headwear" -> EntityModelPartNames.HAT // "hat"
                lower == "body" -> EntityModelPartNames.BODY // "body"
                lower == "jacket" -> EntityModelPartNames.JACKET // "jacket"
                
                // Arms
                lower == "rightarm" || lower == "right_arm" -> EntityModelPartNames.RIGHT_ARM // "right_arm"
                lower == "leftarm" || lower == "left_arm" -> EntityModelPartNames.LEFT_ARM // "left_arm"
                
                // Legs
                lower == "rightleg" || lower == "right_leg" -> EntityModelPartNames.RIGHT_LEG // "right_leg"
                lower == "leftleg" || lower == "left_leg" -> EntityModelPartNames.LEFT_LEG // "left_leg"
                
                // Sleeves/Pants
                lower == "rightsleeve" || lower == "right_sleeve" -> "right_sleeve"
                lower == "leftsleeve" || lower == "left_sleeve" -> "left_sleeve"
                lower == "rightpants" || lower == "right_pants" -> "right_pants"
                lower == "leftpants" || lower == "left_pants" -> "left_pants"
                
                else -> name 
            }
        }
    }
    
   override fun setAngles(state: PlayerEntityRenderState) {
        super.setAngles(state)
        
        if (animationArmsOutFront) {
            setArmAngle(partsMap["rightArm"] ?: partsMap["right_arm"])
            setArmAngle(partsMap["leftArm"] ?: partsMap["left_arm"])
        }

        if (animationStationaryLegs) {
            resetLegAngle("rightLeg", "right_leg")
            resetLegAngle("leftLeg", "left_leg")
        }
    }

    private fun setArmAngle(part: ModelPart?) {
        part?.apply {
            pitch = -1.5707964f
            yaw = 0f
            roll = 0f
        }
    }

    private fun resetLegAngle(key1: String, key2: String) {
        val leg = partsMap[key1] ?: partsMap[key2] ?: return
        val def = defaultTransforms[key1] ?: defaultTransforms[key2] ?: return
        leg.pitch = def.pitch
        leg.yaw = def.yaw
        leg.roll = def.roll
    }
    
    fun copyFromVanilla(vanillaModel: PlayerEntityModel) {
        fun copyRot(bedrockName: String, vanillaPart: ModelPart) {
            val part = partsMap[bedrockName] ?: partsMap[mapBoneName(bedrockName)]
            part?.let {
                it.pitch = vanillaPart.pitch
                it.yaw = vanillaPart.yaw
                it.roll = vanillaPart.roll
            }
        }

        // Copy main parts
        listOf(
            "head" to vanillaModel.head,
            "body" to vanillaModel.body,
            "hat" to vanillaModel.hat
        ).forEach { (name, part) -> copyRot(name, part) }

        if (!animationArmsOutFront) {
            copyRot("rightArm", vanillaModel.rightArm)
            copyRot("leftArm", vanillaModel.leftArm)
        }
        if (!animationStationaryLegs) {
            copyRot("rightLeg", vanillaModel.rightLeg)
            copyRot("leftLeg", vanillaModel.leftLeg)
        }

        // Helper to get pivotY reflectively
        fun getPivotY(part: ModelPart): Float {
            return try {
                val field = part.javaClass.getDeclaredField("pivotY")
                field.isAccessible = true
                (field.get(part) as? Number)?.toFloat() ?: 0f
            } catch (_: Exception) {
                try {
                    val field = net.minecraft.client.model.ModelPart::class.java.getDeclaredField("pivotY")
                    field.isAccessible = true
                    field.getFloat(part)
                } catch (_: Exception) {
                    0f
                }
            }
        }

        // Compute approximate vertical offsets for armor/cape so feature renderers can be translated to match Bedrock geometry
        try {
            val bodyTransform = defaultTransforms["body"] ?: defaultTransforms["BODY"]
            val bedrockBodyY = bodyTransform?.y ?: 0f
            val vanillaBodyPivotY = getPivotY(vanillaModel.body)

            val bedrockHeadY = defaultTransforms["head"]?.y ?: bedrockBodyY
            val vanillaHeadPivotY = getPivotY(vanillaModel.head)

            // Upper armor (helmet + chest) — average head and body deltas
            upperArmorYOffset = ((bedrockBodyY + bedrockHeadY) * 0.5f) - ((vanillaBodyPivotY + vanillaHeadPivotY) * 0.5f)
            armorYOffset = upperArmorYOffset

            // Compute cape and elytra offsets from specific bone pivots if present
            val bedrockCapeY = defaultTransforms["cape"]?.y ?: bedrockBodyY
            capeYOffset = bedrockCapeY - vanillaBodyPivotY

            val bedrockElytraY = defaultTransforms["elytra"]?.y ?: bedrockCapeY
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }
}